Eliminate dublicate inotify events

SeyedMir opened the pull request:
This PR is meant to avoid sending duplicate inotify events from filewatch to the parent process. To this end, we push each event into an abstract table that discards an event if it is already in the table. Having processed a certain number of events, we iterate the abstract table to send out each entry.

The abstract table is currently implemented as a two-level hash table (in TLH_absTab.c). In the first level, we keep track of directory watch descriptors, whereas in the second level, we keep track of filenames in each specific directory. So, in the first level, we have the directory watch descriptor as the key and a (second level) hash table as the value. In the second level, we have the filename as the key and do not have any value.

Currently, glib is used to provide the implementation for each individual hash table.

RAttab commented on c_src/TLH_absTab.c:
```diff
@@ -0,0 +1,225 @@
+     * key points to another (level2) hash table 
+     */
+    GHashTable *lev1; /* directory watch descriptors */
+    GHashTableIter lev1_rmItr, lev2_rmItr; /* used for remove_next_absTab() */
```
I'll defer formatting and styling decision to richard but it does feel weird to see a mix of snake-case and camel-case. I'd suggest picking one and sticking to it.

rkallos replied:
I think it looks ok... Then again, I'm weird cause I love lower kebab-case.

SeyedMir replied:
Agreed. Will update things to be all snake-case.

RAttab commented on c_src/TLH_absTab.c:
```diff
@@ -0,0 +1,225 @@
+     */
+};
+
+AbsTab_t new_absTab()
```
I personally prefer to always place the namespace of my C function first. For this file, I would use the `absTab_xxx` consistently across all exported functions. The advantage of this naming convention are not huge but it does maintain consistency across the entire code base while always placing the object type being manipulated in the same place. In any case, this is down to preference.

SeyedMir replied:
Agreed. Kind of noticed this myself towards the end of my coding!
Will update accordingly.

rkallos commented on c_src/filewatch_inotify.c:
```diff
@@ -186,36 +208,43 @@ static void send_output(ErlDrvData self_)
-        buf[pos++] = (ErlDrvTermData) msg->name;
-        buf[pos++] = strnlen(msg->name, NAME_MAX);
+        buf[pos++] = (ErlDrvTermData) msg.name;
+        buf[pos++] = strnlen(msg.name, NAME_MAX);
```
Shouldn't this use `msg.len` instead of `NAME_MAX`?

SeyedMir replied:
I think NAME_MAX is better because we know that regardless of msg.len, we should never go beyond NAME_MAX.

rkallos replied:
True.

My main worry is that this might accidentally copy extra bytes beyond `msg.name[msg.len]` if there is no null byte at `msg.name[msg.len]`. However, `strncpy` does say it copies a null byte, so maybe I'm just being paranoid.

RAttab commented on c_src/TLH_absTab.c:
```diff
@@ -0,0 +1,225 @@
+    if(msg == NULL) {
+        return 0;
+    }
+    int ret;
```
I recommend to declare variable as close to their use as possible. The strongest rationale for this is that it reduces the number of places in the code where this variable can be potentially used or changed which makes it easier to read and validate the code. In this case, ret is only used to store the result of `g_hash_table_add` so it'd recommend declaring it on that line.

SeyedMir replied:
Glad to hear this because that's how I actually would like to do it myself.

rkallos commented on c_src/TLH_absTab.c:
```diff
@@ -0,0 +1,225 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <glib.h>
```
:fearful:

This might be NIH syndrome, but I have an aversion to including such a huge library for a single data structure. I'm glad that it (filewatch+glib) works on my machine right now, but I would prefer if we either a) (re)used a self-contained hash table library or b) wrote our own specialized hash table.

@gnuvince, you know the most about our pains with adding 3rd-party libs to rtb-gateway & friends, do you care to weigh in?

gnuvince replied:
Certainly, the smaller the better when it comes to preparing binary releases.  I cannot speak on the ease or difficulty of including glib, that's something we'd have test.  If it's written in C and they care about backward compatibility, it could be as simple as adding an extra dependency for the rtb-gateway ebuild.  On the other, if we need to ship the library ourselves, the process would be "make a build, try to run it on a machine, figure out which .so file is currently preventing gateway from starting, add it, and repeat."  In the case of a large library like glib, the number of .so files that we may need to include could be large.

I don't know what people think of [uthash](https://troydhanson.github.io/uthash/), but if it, or another small hashtable implementation, would fit the bill, it could save us much effort down the line.

SeyedMir replied:
Will be working on this.

RAttab replied:
@gnuvince Conversation outside of this PR have suggested using an adapted version of the optics hash table (https://github.com/RAttab/optics/blob/master/src/utils/htable.h) as it's pretty dirt simple and can easily be adapted to our needs here.

RAttab commented on c_src/TLH_absTab.c:
```diff
@@ -0,0 +1,225 @@
+         * we assume this is taken care of by the caller (user) of absTab.
+         */
+    }
+    /* destroy level1 table */
```
This comment is pretty redundant with what the code does. In other words, it's better if your comment focuses on the why instead of the how. Generally speaking you want the code to be self-explanatory which usually involves having good variable and function names.

RAttab commented on c_src/TLH_absTab.c:
```diff
@@ -0,0 +1,225 @@
+    return 0;
+}
+
+int remove_next_absTab(AbsTab_t tab, Message *msg)
```
For iterators like interface, it's best if you store the state of the iterator in an external object. It avoids sharing data between two usage of the iterator which is a potent source of mistakes. So while it's probably fine in this instance, I think it would make the code a bit cleaner.

SeyedMir replied:
I did it this way to avoid explicit use/declaration of iterators in the driver code.

RAttab commented on c_src/absTab.h:
```diff
@@ -0,0 +1,20 @@
+#include <stdio.h>
+
+/* forward declarations */
+typedef struct AbsTab *AbsTab_t;
```
To stay in line with the conventions in this code base I'd avoid typedef-ing away the `struct` keyword. 

This is also fairly subjective and up to personal preference, but I tend to not hide things like pointers in typedefs as it makes the interface less obvious.

RAttab commented on c_src/build.sh:
```diff
@@ -13,8 +13,8 @@ ERTS_INCLUDE_DIR=${ERTS_INCLUDE_DIR:-${ERL_ROOT}/erts-$(erlang_eval 'erlang:syst
 DEFAULT_CFLAGS="-O3 -march=native -mtune=native -ggdb -Wall -Wextra"
-CFLAGS="-fPIC -I${ERTS_INCLUDE_DIR} -I${ERL_INCLUDE_DIR} -std=gnu11 ${CFLAGS:-$DEFAULT_CFLAGS}"
-LDFLAGS="-L${ERL_LIB_DIR} -lei ${LDFLAGS:-}"
+CFLAGS="-fPIC `pkg-config --cflags glib-2.0` -I${ERTS_INCLUDE_DIR} -I${ERL_INCLUDE_DIR} -std=gnu11 ${CFLAGS:-$DEFAULT_CFLAGS}"
```
Consider using the `$()` syntax instead of backticks for command substitution in shell scripts: http://mywiki.wooledge.org/BashFAQ/082

RAttab commented on c_src/filewatch_inotify.c:
```diff
@@ -24,21 +26,16 @@ enum {
     size_t n_events;
     unsigned long cooldown_ms;
-    struct message msgs[MAX_EVENTS];
+    struct Message msgs[MAX_EVENTS];
```
I find it a bit strange that you typedef-ed the `Message` struct yet you still use the `struct` keyword when using the struct. The original reason for typedef-ing structs was that crappy old compiler (**cough** *VS6* **cough**) couldn't handle the struct keyword in variable definitions (later cargo-culted by lazy programmers that wanted to save on 6 letters).

In other words, if you continue to use the `struct` keyword then there's no reason to typedef the struct. Note that if you want to forward declare a struct, you can simply write `struct message;` and that's it.

RAttab commented on c_src/filewatch_inotify.c:
```diff
@@ -71,8 +69,23 @@ static ErlDrvData start(ErlDrvPort port, char *command)
     self->cooldown_ms = cooldown_ms;
 
+    // allocate memory for the name filed in each Message
+    for(i = 0; i < MAX_EVENTS; i++) {
```
Note that C89 (might be C99) has allowed the declaration of variables in for loop headers (eg. `for (int i = 0; ...`) which will be scoped to the loop itself. In other words, I'd recommend scoping the `i` variable to each of the for loops in this function as opposed to reusing the variable.

RAttab commented on c_src/filewatch_inotify.c:
```diff
@@ -241,7 +270,7 @@ static void ready_input(ErlDrvData self_, ErlDrvEvent fd_)
         }
 
-        if (++self->n_events < MAX_EVENTS) {
+        if (self->n_events < MAX_EVENTS) {
```
I have a feeling that something fishy is going on here. At a quick glance of this diff, I can't seem to find where `n_events` is incremented which makes the removal of the `++` suspicious.

RAttab commented on c_src/message.h:
```diff
@@ -0,0 +1,10 @@
+#ifndef MESSAGE_H
+#define MESSAGE_H
```
I recommend the use of `#pragma once` as it's less error prone and less wordy then header guards. We've also standardised around the pragma here as it's more convenient.

RAttab commented on c_src/TLH_absTab.c:
```diff
@@ -0,0 +1,225 @@
+#include "absTab.h"
+#include "message.h"
+
+struct AbsTab {
```
I feel like this struct name doesn't properly reflect what this data structure does. At a conceptual level, what this component tries to do maintain a set of pending notifications so I think the interface could be reworked a little.

There's a couple of approaches that I can think of:
- Treat it like a queue and name it `event_queue` and use verbs like `push`/`pop`. This would be fairly intuitive but would hide the fact that we drop duplicate events.
- Treat it like a set and name it `notification_set` or `event_set` and use verbs like `put`/`next`. It's a bit trickier to come up with the right verbs for the functions on this one but it might be overall better.

RAttab commented:
I did a first pass at the PR but I haven't checked everything (eg. resource ownership) just yet as I expect this to change a lot with the removal of the glib hash table (I really don't like their interface). 

Also note that I forgot to refresh my browser when I resumed my review which means that a bunch of my more recent comments are collapsed so make sure to double-check those as well.

In any case, let me know when you're ready for another round of review :)

rkallos commented:
Also, with these changes, you can remove the call to `ordsets:from_list` in `src/filewatch.erl:109` :smiley:

SeyedMir commented:
I just updated the PR to address the comments so far. In addition, I have added a modified version of the hash table implementation from the optics repo. However, I have not updated the notification_set implementation to use the new hash table (htable) yet; I thought that should be done in a separate step to keep the PR smaller.
Looking forward to receiving your comments!

RAttab commented on c_src/optics_htable/htable.c:
```diff
@@ -0,0 +1,224 @@
+    return NULL;
+}
+
+void htable_itr_init(struct htable *ht, struct htable_iterator *itr)
```
Here's an example of how I use `htable_next` (could be documented a little better): https://github.com/RAttab/optics/blob/master/test/htable_test.c#L119

In other words, I don't think the extra utilities are necessary as the interface already allows for pretty compact iteration but your millage may vary.

SeyedMir replied:
True, it is more compact, but is it better interface-wise? I mean you're using a bucket pointer as an iterator, and you know it works because you know about the details of the htable_next function. But in general, a bucket is a bucket and its implementation could change, in which case, the bucket iterator won't work anymore.

RAttab replied:
If you intend on hiding `struct bucket` from the interface (this is subjective so your call) then I would rework the interface to look something like this:

```
// If this is an implementation detail then don't make it public. 
// Same for `htable_next` which should not be in the header at all.
struct htable_bucket;

// Needs to be defined (not opaque) so it can be stack allocated
// I'm pretty sure these are the only fields that are actually needed but I might be wrong.
struct htable_it { struct htable *ht; struct htable_bucket *bucket; };

bool htable_it_next(struct htable_it *it); // returns false if there are no more elements
struct key * htable_it_key(const struct htable_it *it);
uint64_t htable_it_val(const struct htable_it *it);

struct htable_it htable_begin(struct htable *ht);
```
This version works best with a while loop. Adding an `htable_it_end` function and having `htable_it_next` return `void` would make it work well with for-loops. Tweak at your discretion.

SeyedMir replied:
What do you think about putting the key and value (pointers) in the iterator and remove htable_it_key and htable_it_val functions?

RAttab replied:
I  feel that you're just back to having the old `struct htable_bucket` implementation but with an extra layer of indirection. Note that I usually tweak an interface as I'm using it to suite my needs so feel free to tweak as you see fit.

Note that my original version was used extensively and has worked quite well for me in the past :)

RAttab commented on c_src/optics_htable/htable.c:
```diff
@@ -0,0 +1,224 @@
+
+void htable_itr_remove(struct htable_iterator *itr)
+{
+    free((struct Key *) itr->bucket->key);
```
You should probably make sure that the key is not null before you try to remove it (the free won't fail but the length decrement would screw things up).

RAttab commented on c_src/optics_htable/htable.h:
```diff
@@ -0,0 +1,88 @@
+void htable_reset(struct htable *);
+void htable_reserve(struct htable *, size_t items);
+struct Key *make_key_int(int i);
+struct Key *make_key_str(char *str);
```
I'd tweak this interface a little bit so that it looks something like this (minus the comments):
```
// opaque type since there's no need to expose the internals of this struct to the user and it's always heap allocated.
struct key; 

// make the number of available bits explicit and add the htable namespace to the function name.
struct key * htable_key_int(uint64_t i); 

// Pointer should be const since we're not modifying it
struct key * htable_key_str(const char * str); 

// I always pair allocation functions with a free function
void htable_key_free(struct key * key);
```

RAttab commented on c_src/optics_htable/htable.h:
```diff
@@ -0,0 +1,88 @@
+// struct
+// -----------------------------------------------------------------------------
+
+enum { htable_key_max_len = 1024 };
```
This restriction is not really relevant for your use case and should really be `NAME_MAX` or whatever that define is.

RAttab commented on c_src/optics_htable/htable_test.c:
```diff
@@ -0,0 +1,130 @@
+#include <stdio.h>
```
Were you aware of the optics htable tests and (very simple) benchmarks?
https://github.com/RAttab/optics/blob/master/test/htable_test.c
https://github.com/RAttab/optics/blob/master/test/htable_bench.c

SeyedMir replied:
Saw htable_test but did not know about htable_bench. I'll make a modified version of them

RAttab replied:
the benches are going to be considerably more complicated so I wouldn't worry too much about those.

SeyedMir replied:
OK, noted

SeyedMir commented:
I updated the PR which 
1) addresses the comments (those relevant after changes), 
2) updates notification_set (now notif_set) implementation based on htable, and
3) updates filewatch_inotify.c based on the new changes.

Looking forward to receiving your comments.

RAttab commented on c_src/filewatch_inotify.c:
```diff
@@ -181,41 +182,57 @@ static void send_output(ErlDrvData self_)
-    strncpy(msg->name, evt->name, NAME_MAX);
+    //Insert into notif_set to keep track of unique {wd, name} tuples.
+    notif_set_insert(self->notif_set,
+                            &(struct Message) {.wd = evt->wd, .name = (char *) evt->name});
```
Since this is the only place that `notif_set_insert` is called I would just pass `wd` and `name` directly instead of wrapping them in a temporary object. As it stands, the first time I read this code I immediately had concerns over the ownership of this temporary object.

RAttab commented on c_src/message.h:
```diff
@@ -0,0 +1,8 @@
+#pragma once
+#include <limits.h>
+
+struct Message {
```
I would just incorporate this struct into the `notif_set` header as it currently exists mostly for interaction with the `notif_set` structure.

RAttab commented on c_src/notif_set.h:
```diff
@@ -0,0 +1,20 @@
+
+struct Notif_set *notif_set_new();
+int notif_set_destroy(struct Notif_set *tab);
+int notif_set_insert(struct Notif_set *tab, struct Message *msg);
```
When passing in pointer params, consider making them `const` if you the function doesn't change them. In this case, I originally was concerned that you passed a temporary object to a function that was planning on modifying it.

RAttab commented on c_src/notif_set_tlh_glib.c:
```diff
@@ -0,0 +1,215 @@
+#include <stdio.h>
```
Any reason why we're keeping the glib version around?

RAttab commented on c_src/notif_set_tlh_htable.c:
```diff
@@ -0,0 +1,198 @@
+
+struct Notif_set *notif_set_new()
+{
+    struct Notif_set *tlh = malloc(sizeof(struct Notif_set));
```
I'd recommend using the following: `struct Notif_set *tlh = malloc(sizeof(*tlh));`
Using the size of the variable instead of the size of the type is less error prone (eg. changed the type on the variable declaration but forgot to change the `sizeof` call).

RAttab replied:
Also, what does `tlh` stand for? I can't quite figure it out which probably means that a better name is warranted.

RAttab replied:
Also, I'd recommend allocating like so instead: `calloc(1, sizeof(*tlh));`
This will zero-initialise your entire structure which avoid having uninitialised fields.

SeyedMir replied:
tlh stands for two-level hash; three-letter acronym :)

RAttab commented on c_src/notif_set_tlh_htable.c:
```diff
@@ -0,0 +1,198 @@
+    return tlh;
+}
+
+int notif_set_destroy(struct Notif_set *tab)
```
I'd recommend using `bool` instead of `int` along with the `true` and `false` return values. This narrows down the set of possible return values which makes it easier for the user of the function to figure out how to handle the return values. Note that I'd even go further and make it `void` since it can only return `true` at the moment.

RAttab replied:
This applies to most functions in this class so I won't repeat myself.

RAttab commented on c_src/notif_set_tlh_htable.c:
```diff
@@ -0,0 +1,198 @@
+
+    // add the filename to the level2 hash table of this wd
+    key = htable_key_str(msg->name);
+    if(!htable_put(lev2_tab, key, 0).ok) { // TODO: store len as well
```
I believe the `TODO` comment here is no longer relevant as the key contains it's own len.

RAttab commented on c_src/notif_set_tlh_htable.c:
```diff
@@ -0,0 +1,198 @@
+        // We have a new directory wd; insert to level1
+        // First create a level2 table for this wd (the value)
+        lev2_tab = malloc(sizeof(struct htable));
+        assert(lev2_tab != NULL);
```
Here you're asserting on the return value of the malloc call while in the `notif_set_new` function you're returning an error response instead. I being consistent in the error handling and in this case, returning an error instead of asserting.

RAttab commented on c_src/notif_set_tlh_glib.c:
```diff
@@ -0,0 +1,215 @@
+
+int notif_set_find(struct Notif_set *tab, struct Message *msg, void *val)
+{
+    (void) val; //To eliminate unused variable warning
```
Why the extra parameter if it's not used?

RAttab replied:
Crap reading the won't file... nevermind...

RAttab commented on c_src/notif_set_tlh_glib.c:
```diff
@@ -0,0 +1,215 @@
+int notif_set_del(struct Notif_set *tab, struct Message *msg)
+{
+    if(msg == NULL) {
+        return 0;
```
I would consider passing a `NULL` as the key to delete to be a programmer error and should therefor `abort`. In other words this is a breach of the `notif_set_del` contract and therefore should fail hard.

RAttab replied:
Reading the wrong file... nevermind...

RAttab commented on c_src/notif_set_tlh_htable.c:
```diff
@@ -0,0 +1,198 @@
+{
+    // IMPORTANT: Each htable owns its keys, but the 
+    //            values are owned by the user.
+    if(msg == NULL) {
```
I would consider passing a NULL as the key to delete to be a programmer error and should therefor abort. In other words this is a breach of the notif_set_del contract and therefore should fail hard.

RAttab commented on c_src/notif_set_tlh_htable.c:
```diff
@@ -0,0 +1,198 @@
+
+int notif_set_find(struct Notif_set *tab, struct Message *msg, void *val)
+{
+    (void) val; // To eliminate unused variable warning
```
Why the extra parameter if it's never used?

RAttab commented on c_src/notif_set_tlh_htable.c:
```diff
@@ -0,0 +1,198 @@
+
+struct Notif_set_itr *notif_set_itr_init(struct Notif_set *tab)
+{
+    struct Notif_set_itr *itr = calloc(1, sizeof(*itr));
```
Note that since this object is small I would just return it instead of allocating it on the heap. This would save on an allocation and removes the need to ensure that it's been freed properly.

SeyedMir replied:
But then I have to make the iterator non-opaque, which is something that I'm reluctant to do.

RAttab commented on c_src/notif_set_tlh_htable.c:
```diff
@@ -0,0 +1,198 @@
+    // Advance the second-level iterator first
+    if((itr->itr2 = htable_next((struct htable *) itr->itr1->value, itr->itr2))) {
+        assert(itr->itr1->key->n == sizeof(int));
+        msg->wd = *((int*) itr->itr1->key->bytes); //Really don't like this
```
I'd recommend creating a function that takes a key and returns the proper type of it. Will give you a chance of doing a bit of validation.

SeyedMir replied:
I am not sure how this can be done correctly. Do you mean checking the number of bytes and deciding based on that?

RAttab commented on c_src/notif_set_tlh_htable.c:
```diff
@@ -0,0 +1,198 @@
+
+size_t notif_set_get_length(struct Notif_set *tab)
+{
+    return (tab != NULL) ? tab->len : 0;
```
I would error out on `NULL` and it's almost certainly a programmer mistake.

RAttab commented on c_src/optics_htable/htable.c:
```diff
@@ -0,0 +1,230 @@
+/* htable.c
+   Rémi Attab (remi.attab@gmail.com), 10 Mar 2016
+   FreeBSD-style copyright and disclaimer apply
+*/
```
I'd also mentioned that it's been modified for the use cases of this library.

RAttab commented on c_src/optics_htable/htable.c:
```diff
@@ -0,0 +1,230 @@
+{
+    struct Key *key = malloc(sizeof(struct Key) + sizeof(int));
+    assert(key != NULL);
+    key->n = sizeof(int);
```
your function is accepting an `uint64_t` as a parameter which is not consistent with the body of the function which is handling `int`s.

RAttab commented:
Finished another round of reviews.

SeyedMir commented:
Thank you very much for your comments, Remi. I'll update things accordingly.

rkallos commented on c_src/build.sh:
```diff
@@ -37,6 +37,7 @@ fi
 
 TARGET=${TARGET:-./priv/filewatch.so}
 SRC=./c_src
+SRCFILES="$SRC/$IMPLEMENTATION.c $SRC/notif_set_tlh_htable.c $SRC/optics_htable/htable.c"
```
To my knowledge, we don't need to compile the hash-table code for the filewatch_noop implementation, so it might be easier to set SRCFILES in the `case "$OS"` section.

rkallos commented on c_src/filewatch_inotify.c:
```diff
@@ -181,41 +185,65 @@ static void send_output(ErlDrvData self_)
+        buf[pos++] = strnlen(msg[num_entries_removed].name, NAME_MAX);
         buf[pos++] = ERL_DRV_TUPLE;
         buf[pos++] = BUFFER_TUPLE_SIZE;
+        
```
Weird that GH isn't picking up this trailing whitespace... It's usually pretty good about this.

Make sure you `:%s/\s\+$//e` at some point :smile_cat:

rkallos commented on src/filewatch.erl:
```diff
@@ -90,12 +90,13 @@ create_watch_map([{Path, Term} | Rest], Port, Map) ->
 
 add_to_watch_map(Dir, File, Term, Port, Map) ->
-    {ok, WatchDescriptor} =
+    {WatchDescriptor, NewMap} = 
```
Some more trailing whitespace

rkallos commented on c_src/notif_set_tlh_htable.c:
```diff
@@ -0,0 +1,245 @@
+        if(htable_key_free((struct Key*) itr->itr2->key)) {
+            itr->itr2->key = NULL;
+            ((struct htable*)(itr->itr1->value))->len--;
+            // TODO: We could destroy the level2 htable and remove
```
This is a good justification for not destroying the level2 table here. I would leave the comment, but remove the TODO.

rkallos commented on c_src/filewatch_inotify.c:
```diff
@@ -181,41 +185,65 @@ static void send_output(ErlDrvData self_)
+    //Insert into notif_set to keep track of unique {wd, name} tuples.
+    int err_flag;
+    notif_set_insert(self->notif_set,
+                         &(struct Message) {.wd = evt->wd, .name = (char *) evt->name},
```
It seems your usual style is to indent to the level of the open paren. I take it this is an artifact of when there the function name was a bit longer? :stuck_out_tongue:

SeyedMir replied:
You're right about my usual style; seems like I missed this one :)

rkallos commented on c_src/notif_set_tlh_htable.c:
```diff
@@ -0,0 +1,245 @@
+    CHECK_SUCCESS(key);
+    if(htable_del((struct htable*) get_ret.value, key).ok) {
+        tab->len--;
+        // TODO: We could destroy the level2 htable and remove
```
Again, it might be good to remove the TODO here.

SeyedMir replied:
Thank you very much for your comments, Richard. I updated things accordingly.

rkallos commented:
The squashed commits look much better! Thanks!

When compiling the source code I receive the following warnings:
```./c_src/notif_set_tlh_htable.c: In function ‘notif_set_itr_next’:
./c_src/notif_set_tlh_htable.c:148:9: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
         msg->wd = *((uint64_t*) itr->itr1->key->bytes); //Really don't like this
         ^~~
./c_src/notif_set_tlh_htable.c: In function ‘notif_set_fprint’:
./c_src/notif_set_tlh_htable.c:196:21: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
                     *((int*) itr1->key->bytes),
                     ^
```

Is there an easy way to fix this?

RAttab commented:
Either `-fno-strict-aliasing` or type punning through union (https://github.com/RAttab/optics/blob/master/src/utils/type_pun.h) which is the only well defined method for type-punning on gcc.

SeyedMir commented:
I can change it to use the union-based type punning if we should not cancel strict aliasing. So, please just let me know. I also wonder why I did not get the same warning on my system; I'll double-check it tomorrow.
